internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -143,132 +141,146 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
-	if preReconcileInfrastructureReady != cluster.Status.InfrastructureReady {
-		r.recorder.Eventf(cluster, corev1.EventTypeNormal, "InfrastructureReady", "Cluster %s InfrastructureReady is now %t", cluster.Name, cluster.Status.InfrastructureReady)
+	if ready && !cluster.Status.InfrastructureReady {
+		log.Info("Infrastructure provider has completed cluster infrastructure provisioning and reports status.ready", cluster.Spec.InfrastructureRef.Kind, klog.KObj(s.infraCluster))
 	}
 
 	// Report a summary of current status of the infrastructure object defined for this cluster.
 	conditions.SetMirror(cluster, clusterv1.InfrastructureReadyCondition,
-		conditions.UnstructuredGetter(infraConfig),
+		conditions.UnstructuredGetter(s.infraCluster),
 		conditions.WithFallbackValue(ready, clusterv1.WaitingForInfrastructureFallbackReason, clusterv1.ConditionSeverityInfo, ""),
 	)
 
-	if !ready {
+	// There's no need to go any further if the Cluster is marked for deletion.
+	if !s.infraCluster.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// If the infrastructure provider is not ready (and it wasn't ready before), return early.
+	if !ready && !cluster.Status.InfrastructureReady {
 		log.V(3).Info("Infrastructure provider is not ready yet")
 		return ctrl.Result{}, nil
 	}
 
 	// Get and parse Spec.ControlPlaneEndpoint field from the infrastructure provider.
 	if !cluster.Spec.ControlPlaneEndpoint.IsValid() {
-		if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Spec.ControlPlaneEndpoint, "spec", "controlPlaneEndpoint"); err != nil && err != util.ErrUnstructuredFieldNotFound {
+		if err := util.UnstructuredUnmarshalField(s.infraCluster, &cluster.Spec.ControlPlaneEndpoint, "spec", "controlPlaneEndpoint"); err != nil && err != util.ErrUnstructuredFieldNotFound {
 			return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
 				cluster.Name, cluster.Namespace)
 		}
 	}
 
 	// Get and parse Status.FailureDomains from the infrastructure provider.
 	failureDomains := clusterv1.FailureDomains{}
-	if err := util.UnstructuredUnmarshalField(infraConfig, &failureDomains, "status", "failureDomains"); err != nil && err != util.ErrUnstructuredFieldNotFound {
+	if err := util.UnstructuredUnmarshalField(s.infraCluster, &failureDomains, "status", "failureDomains"); err != nil && err != util.ErrUnstructuredFieldNotFound {
 		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve Status.FailureDomains from infrastructure provider for Cluster %q in namespace %q",
 			cluster.Name, cluster.Namespace)
 	}
 	cluster.Status.FailureDomains = failureDomains
 
+	// Only record the event if the status has changed
+	if !cluster.Status.InfrastructureReady {
+		r.recorder.Eventf(cluster, corev1.EventTypeNormal, "InfrastructureReady", "Cluster %s InfrastructureReady is now %t", cluster.Name, cluster.Status.InfrastructureReady)
```

Comment:
Using  cluster.Status.InfrastructureReady in the message seems wrong (it should be false in this case??)